cpp/bench/ann/src/common/benchmark.hpp

```diff
@@ -180,129 +189,168 @@ void bench_search(::benchmark::State& state,
+      } catch (const std::exception& ex) {
+        state.SkipWithError("The algorithm '" + index.name +
+                            "' requires the base set, but it's not available. " +
+                            "Exception: " + std::string(ex.what()));
+        return;
+      }
     }
   }
 
-  std::ptrdiff_t batch_offset   = 0;
-  std::size_t queries_processed = 0;
+  const auto algo_property = *current_algo_props;
+  const T* query_set       = dataset->query_set(algo_property.query_memory_type);
+
+  /**
+   * Each thread will manage its own outputs
+   */
+  std::shared_ptr<buf<float>> distances =
+    std::make_shared<buf<float>>(algo_property.query_memory_type, k * query_set_size);
+  std::shared_ptr<buf<std::size_t>> neighbors =
+    std::make_shared<buf<std::size_t>>(algo_property.query_memory_type, k * query_set_size);
+
+  auto start = std::chrono::high_resolution_clock::now();
   cuda_timer gpu_timer;
   {
     nvtx_case nvtx{state.name()};
+
+    // TODO: Have the odd threads load the queries backwards just to rule out caching.
+    ANN<T>* algo = dynamic_cast<ANN<T>*>(current_algo.get());
     for (auto _ : state) {
-      // measure the GPU time using the RAII helper
       [[maybe_unused]] auto ntx_lap = nvtx.lap();
       [[maybe_unused]] auto gpu_lap = gpu_timer.lap();
+
+      auto start = std::chrono::high_resolution_clock::now();
       // run the search
       try {
         algo->search(query_set + batch_offset * dataset->dim(),
                      n_queries,
                      k,
-                     neighbors.data + batch_offset * k,
-                     distances.data + batch_offset * k,
+                     neighbors->data + batch_offset * k,
+                     distances->data + batch_offset * k,
                      gpu_timer.stream());
       } catch (const std::exception& e) {
         state.SkipWithError(std::string(e.what()));
       }
+
+      auto end = std::chrono::high_resolution_clock::now();
+
+      auto elapsed_seconds = std::chrono::duration_cast<std::chrono::duration<double>>(end - start);
       // advance to the next batch
       batch_offset = (batch_offset + n_queries) % query_set_size;
       queries_processed += n_queries;
+      state.SetIterationTime(elapsed_seconds.count());
+      total_time += elapsed_seconds.count();
     }
   }
+  auto end = std::chrono::high_resolution_clock::now();
+  if (state.thread_index() == 0) {
+    auto duration = std::chrono::duration_cast<std::chrono::duration<double>>(end - start).count();
+    state.counters.insert({{"end_to_end", duration}});
+  }
   state.SetItemsProcessed(queries_processed);
-  state.counters.insert({{"k", k}, {"n_queries", n_queries}});
   if (cudart.found()) {
-    state.counters.insert({{"GPU Time", gpu_timer.total_time() / state.iterations()},
-                           {"GPU QPS", queries_processed / gpu_timer.total_time()}});
+    state.counters.insert({{"GPU", gpu_timer.total_time() / double(state.iterations())}});
```

Comment:
I would keep it as "GPU Time" or "GPU Latency", since that is more descriptive. 
```suggestion
  double gpu_latency_per_iteration = gpu_timer.total_time() / double(state.iterations());
  state.counters.insert({"GPU Latency", {gpu_latency_per_iteration, benchmark::Counter::kAvgThreads}});
```

Reply:
I think we shall look into the `gpu_timer` as a separate issue. What is the expectation from the `ANN::search()` methods that we call within the benchmark loop, are they expected to sync? Currently they do, but  I would like to review it later because different streams and events are synced.

Reply:
Yes, the "Does that even make sense?" question also occurred to me and I suspect the answer is no (but @achirkin might prove me wrong).

Reply:
Thanks Artem for the comment! I agree with the three points that motivated the design, and I also confirm that the GPU time that we measure agrees with the cpu wall time.

What is puzzling to me: why the `cuda_lap` timer achieves this goal? https://github.com/rapidsai/raft/blob/9ad76faeec798a45c359fb358c414f1d3a19eb3f/cpp/bench/ann/src/common/util.hpp#L108-L127

We create a new `stream_`, register a `start_` event before we start the actual work. For concreteness let's assume we call `ivf_pq::search()` in the benchmark loop. The raft_ivf_pq_wrapper class will create a new `device_resources` object and will use `rmm::cuda_stream_per_thread` to schedule the work. In the end we add a `stop_` event, sync with it and measure duration. Since there is no work queed on `stream_`, what guarantees that `ivf_pq::search() `falls between `start_` and `stop_`?

Reply:
From my understanding and the behavior I've seen, a new gpu_timer stream is created for each thread and the call to search() for each RAFT index type is synchronous wrt it's thread, thus the gpu_timer will wait before for the search to complete before each lap, and thus that wait time will cause the same time gap between start and end events.

Im not sure the faiss indexes will function this way, as I believe they are syncing on the stream that's passed into search() (plus any additional syncs that happen to the stream in the corresponding StandardGouResources instance. We should be able to set the stream on that instance, so we could very well set the cuda stream per thread on it, I believe, which would make it function more like raft.

Reply:
I think the assumption here is that we will have a` cudaStreamSync()` or equivalent while calling `ivf_pq::search()`. That would block the CPU thread, and delay scheduling the `stop_` event.

Reply:
Let's continue the discussion here: https://github.com/rapidsai/raft/pull/1940/files#r1376957640

Reply:
> I think the assumption here is that we will have a cudaStreamSync() or equivalent while calling ivf_pq::search(). That would block the CPU thread, and delay scheduling the stop_ event.

This should be the case, though, shouldn't it?

Reply:
The idea behind `cuda_lap` is to not require any extra synchronization, such as `cudaStreamSync()`. The only requirement is that ivf_pq::search() does something in the stream passed to it.
What `cuda_lap` does:
1. Waits till previous work in `stream_` finishes and submits the `start_` event into it. A somewhat risky assumption here is that the event triggers immediately. Maybe, in reality it could trigger some time between the submission and the first use of stream_ (which could be a synchronization or a data transfer or a kernel)? To fix it, we can just add another cudaEventSynchronize or cudaStreamSynchronize right after submitting the event (but I'm not sure this need fixing).
2. Submits the `stop_` event into the `stream_` after the call to the implementation and immediately synchronizes on it. Hence it captures everything the implementation has done in the `stream_`. However, since the implementation is called in the same thread, the `stop_` event is submitted no earlier than the CPU side of the algorithm finishes. Hence, the `stop_` time is the larger of the two (cpu completion time, gpu completion time). 

As a result, the two events together capture the total execution time, including the GPU work that could still be happening in the `stream_` after the CPU side has already finished. Yet, the GPU time cannot be larger than our recorded wall time, because we synchronize on the `stop_` event within the benchmark iteration (in the destructor of the `cuda_lap`).

The implementation must mark its work in the `stream_`, but not necessarily via synchronizing with the host.
Currently, both FAISS and raft_ivf_pq wrappers have their own streams created as a part of their states. That is, we don't create a new raft_resources handle with this stream and don't set this as the main stream in FAISS! This is done to make sure we can cache the algo handles between runs without using already destroyed streams. In both cases, I use `cudaEventRecord` to establish a dependency between the internal implementation streams and the passed cuda_lap stream.